File: Array/009-PascalTriangle2/PascalTriangle2.py
"""
Given an integer rowIndex, return the rowIndexth (0-indexed) row of the Pascal's triangle.
In Pascal's triangle, each number is the sum of the two numbers directly above it as shown:

Example :
Input: rowIndex = 3
Output: [1,3,3,1]
"""

class Solution(object):
    def getRow(self, rowIndex):
        lst = [[1]]
        for i in range(1,rowIndex+1):
            lst += [map(lambda x,y: x+y, [0]+lst[-1], lst[-1]+[0])]
        return lst[rowIndex]
      
"""
This problem is basically the same as pascal triangle I problem. We can use optimal solution from the last problem (adding offset of list to create next list)
and then returning only the wanted list.
"""

File: Array/009-PascalTriangle2/PascalTriangle2.go
func getRow(rowIndex int) []int {
    out := make([][]int, rowIndex+1)
    out[0] = []int{1}
    
    for i:=1; i<rowIndex+1; i++ {
        prev := out[i-1]
        temp := make([]int, i+1)
        temp[0] = 1
        temp[i] = 1
        
        for l:=1; l<i; l++ {
            temp[l] = prev[l] + prev[l-1]
        } 
        
        out[i] = temp
    }
    
    return out[len(out)-1]
}
